src/openrct2/peep/Peep.cpp

```diff
@@ -2673,14 +2673,141 @@ static void peep_go_to_ride_entrance(rct_peep * peep, Ride * ride)
+
+        uint8 num_seats = vehicle->num_seats;
+        if (vehicle_is_used_in_pairs(vehicle))
+        {
+            num_seats &= VEHICLE_SEAT_NUM_MASK;
+            if (vehicle->next_free_seat & 1)
+            {
+                peep->current_car = i;
+                peep_choose_seat_from_car(peep, ride, vehicle);
+                peep_go_to_ride_entrance(peep, ride);
+                return false;
+            }
+        }
+        if (num_seats == vehicle->next_free_seat)
+            continue;
+
+        if (ride->mode == RIDE_MODE_FORWARD_ROTATION || ride->mode == RIDE_MODE_BACKWARD_ROTATION)
+        {
+            uint8 position = (((~vehicle->vehicle_sprite_type + 1) >> 3) & 0xF) * 2;
+            if (vehicle->peep[position] != SPRITE_INDEX_NULL)
+                continue;
+        }
+
+        *car_array_pointer++ = i;
+    }
+
+    *car_array_size = (uint8)(car_array_pointer - car_array);
+
+    if (*car_array_size == 0)
+        return false;
+
+    return true;
+}
+
+static void peep_update_ride_ss_at_entrance_try_leave(rct_peep * peep)
+{
+    // Destination Tolerance is zero when peep has completely
+    // entered entrance
+    if (peep->destination_tolerance == 0)
+    {
+        remove_peep_from_queue(peep);
+        peep_decrement_num_riders(peep);
+        peep->state = PEEP_STATE_FALLING;
+        peep_window_state_update(peep);
+    }
+}
+
+static bool peep_check_ride_price_at_entrance(rct_peep * peep, Ride * ride, money32 ridePrice)
+{
+    if ((peep->item_standard_flags & PEEP_ITEM_VOUCHER) &&
+        peep->voucher_type == VOUCHER_TYPE_RIDE_FREE &&
+        peep->voucher_arguments == peep->current_ride)
+        return true;
+    
+    if (peep->cash_in_pocket <= 0)
+    {
+        peep_insert_new_thought(peep, PEEP_THOUGHT_TYPE_SPENT_MONEY, PEEP_THOUGHT_ITEM_NONE);
+        peep_update_ride_ss_at_entrance_try_leave(peep);
+        return false;
+    }
+
+    if (ridePrice > peep->cash_in_pocket)
+    {
+        peep_insert_new_thought(peep, PEEP_THOUGHT_TYPE_CANT_AFFORD_0, peep->current_ride);
+        peep_update_ride_ss_at_entrance_try_leave(peep);
+        return false;
+    }
+
+    uint16 value = ride->value;
+    if (value != RIDE_VALUE_UNDEFINED)
+    {
+        if (value * 2 < ridePrice)
+        {
+            peep_insert_new_thought(peep, PEEP_THOUGHT_TYPE_BAD_VALUE, peep->current_ride);
+            peep_update_ride_ss_at_entrance_try_leave(peep);
+            return false;
+        }
+    }
+    return true;
+}
+
 /**
  *
  *  rct2: 0x00691A3B
  */
-static void peep_update_ride_sub_state_0(rct_peep * peep)
+static void peep_update_ride_ss_at_entrance(rct_peep * peep)
 {
     Ride * ride = get_ride(peep->current_ride);
 
+    // The peep will keep advancing in the entranceway
+    // whilst in this state. When it has reached the very
+    // front of the queue destination tolerance is set to 
+    // zero to indicate it is final decision time (try_leave will pass).
```

Comment:
This state is also used for guests that exit though the entrance back to the front of the queue, which may be nice to clarify here.